base/compiler/ssair/passes.jl

```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
     return complete(compact)
 end
 
+function arrayptr(P::Union{Type{Ptr{T}},Type{Ptr{Cvoid}}}, x::Array{T}, i::Int)::P where T
+    # Based on unsafe_convert for RefArray
+    if allocatedinline(T)
+        p = pointer(x, i)
+    elseif isconcretetype(T) && ismutabletype(T)
+        p = pointer_from_objref(x[i])
+    else
+        p = pointerref(Ptr{Ptr{Cvoid}}(pointer(x, i)), 1, Core.sizeof(Ptr{Cvoid}))
+    end
+    return p
+end
+
+struct CongruenceClass
+    ssa::Int
+    blockidx::Int
+end
+
+function perform_symbolic_evaluation(stmt::Expr, ssa_to_ssa)
+    # rename all SSAValues
+    # taken from renumber_ir_elements!
+    if stmt.head !== :enter && !is_meta_expr_head(stmt.head)
+
+        # key = similar(stmt.args, length(stmt.args)+1)
+        # copyto!(key, stmt.args)
+        # key[end] = stmt.head
+        # for (i, arg) in enumerate(stmt.args)
+        #     if isa(arg, SSAValue)
+        #         key[i] = SSAValue(ssa_to_ssa[arg.id])
+        #     end
+        # end
+        # svec(key...)
+
+        key = ccall(:jl_alloc_svec, SimpleVector, (Csize_t,), length(stmt.args)+1);
```

Comment:
Clever, but no. That said, since the only thing this is needed for is congruences, it would be fine to do this with an array, but you'd need to make sure to root all Expr args arrays for the lifetime of the computation, which is possible, but I don't think it's worth the tradeoff. I don't see why the Vector{Any} version above should be that much slower.

Reply:
Also, I would be fine with doing this optimization in the allocation optimization passes in the backend, it's just not the right place to hardcode here.

Reply:
That said, this is also of course just using an svec allocation to compute a hash key. It should be just as possible to just compute the hash key directly without the extra allocation.

Reply:
The performance hit is due to having to allocate both a Vector and a svec and then copying the Vector to the svec.  `gvn!` on `Base.code_ircode(exp, (Float64,); optimize_until="compact 2"` using the Vector{Any} implementation takes 36ms, here's the profile of this specific code
![image](https://github.com/JuliaLang/julia/assets/19725290/ecdfabbc-155d-41ff-9b1c-75f96b0ce453)
And using the unsafe svec implementation 27ms, 
![image](https://github.com/JuliaLang/julia/assets/19725290/f37b92a4-523e-4932-92a6-cabd26a29db5)